File: python/abc352/d.py
from atcoder.segtree import SegTree

n,k = map(int, input().split())
p = list(map(int, input().split()))

dict_p = {p[i]:i+1 for i in range(n)}
sorted_p = [dict_p[i] for i in range(1,n+1)]

seg_tree_max = SegTree(op = max, e = 0, v = sorted_p)
seg_tree_min = SegTree(op = min, e = n, v = sorted_p)

min_value = n
for i in range(n-k+1):
    i_k = seg_tree_max.prod(i,i+k)
    i_1 = seg_tree_min.prod(i,i+k)
    min_value = min(min_value, i_k-i_1)

print(min_value)

File: rust/abc352/src/d.rs
#[allow(unused_imports)]
use proconio::{
    fastout, input,
};
use std::collections::HashMap;
use std::cmp::min;

#[fastout]
fn main() {
	input! {
        n: usize,
        k: usize,
        p: [usize; n],
    }

    let mut dict_p = HashMap::new();
    for i in 0..n {
        dict_p.insert(p[i], i+1);
    }

    let mut sorted_p = vec![0; n];
    for i in 1..=n {
        sorted_p[i-1] = dict_p[&i];
    }

    let mut ans = n;
    for i in 0..=n-k {
        let cloned_partially_sorted_p = Vec::from_iter(&mut sorted_p[i..i+k].iter().cloned());
        let i_k = cloned_partially_sorted_p.iter().max().unwrap();
        let i_1 = cloned_partially_sorted_p.iter().min().unwrap();
        ans = min(ans, i_k - i_1);
    }

    println!("{}", ans);
}
